lib/msf/ui/debug.rb

```diff
@@ -0,0 +1,171 @@
+
+        if @@driver.active_module
+          ds = @@driver.active_module.datastore.dup
+
+          # Ensures that the local default value of a variable isn't set if a global value of a variable is set
+          ds.keys.each do |k|
+            # Removes an active module datastore item if it has a default value & it matches the default value
+            if @@driver.active_module.options.key?(k.upcase) && @@driver.active_module.options[k.upcase].default == ds[k]
+              ds.delete(k)
+            end
+          end
+
+          add_hash_to_ini_group(ini, ds, @@driver.active_module.refname)
+        end
+
+        banner = build_banner('Datastore')
+
+        if ini.to_s.empty?
+          banner + "The local config file is empty, no global variables are set, and there is no active module.\n\n"
+        else
+          banner + ini.to_s
+        end
+      end
+
+      def self.get_errors
+        errors = File.read(Msf::Config.log_directory + File::SEPARATOR + "error.log")
+
+        #Returns the errors in error.log file as an array | Separator of errors is two consecutive \n chars
+        res = errors.scan(/(\[\d\d\/\d\d\/\d\d\d\d \d\d:\d\d:\d\d\] \[.*?\] error:(?:(?!\n\n).)+)/m)
+
+        #Scan returns each error as a single item array
+        res.flatten!
+        #Scan reverses initial LIFO order of +errors+
+        res.reverse!
+
+        build_banner('Errors') + concat_str_array_from_last_idx(res, ERROR_TOTAL, true)
+      end
+
+      def self.get_logs
+        log_lines = File.readlines(Msf::Config.log_directory + File::SEPARATOR + "framework.log")
+
+        build_banner('Logs') + concat_str_array_from_last_idx(log_lines, LOG_LINE_TOTAL)
+      end
+
+      def self.get_versions
+        major, minor, patch = @@framework.version.split('-')
+
+        str = ""
+        str += "Framework Major Version: #{major} \n"
+        str += "Framework Minor Version: #{minor} \n"
+        str += "Framework Patch Version: #{patch} \n"
+        str += "Ruby Version: #{RUBY_DESCRIPTION} \n"
+
+        build_banner('Versions') + str
+      end
+
+      #######
+      private
+      #######
+
+      def self.add_hash_to_ini_group(ini, hash, group_name)
+        if hash.empty?
+          return
+        end
+
+        unless ini.group?(group_name)
+          ini.add_group(group_name)
+        end
+
+        hash.each_pair do |k,v|
+          ini[group_name][k] = v
+        end
+      end
+
+      def self.concat_str_array_from_last_idx(array, concat_total, extra_padding=false)
+        end_pos = array.length > concat_total ? array.length - concat_total : 0
+        start_pos = array.length - 1
+
+        str = ""
+        while start_pos >= end_pos
+          str += array[start_pos]
+          str += "\n\n" if extra_padding
+          start_pos -= 1
+        end
+
+        str.strip + "\n\n"
+      end
+
+      def self.build_banner(banner_name)
+        str =  "###########################################\n"
```

Comment:
The end result is markdown, so a header is formatted as:

```
def self.build_banner(banner_name)
  "##  %grn#{banner_name}%clr\n"
end
```

Reply:
Added, will also be adding in markdown to match https://wiki.corp.rapid7.com/display/MSF/Improved+Error+Reporting